### PR TITLE
[project-pineapples <-] full async request router

### PR DESCRIPTION
**Pull request by spacejam**
This PR adds support for driving concurrent requests from the RequestRouter out-of-order, allowing things like redirections and retries to happen immediately upon a specific future being completed due to a timeout or redirection, without any head-of-line blocking when driving all requests concurrently. The basic idea is to have our `Promise` support a "Fill notifier" that is called when `Promise::Fill` is called. In this case, the RequestRouter has a `Notifier` that it shares with the transport interface through this Fill-notifier, and after sending out concurrent requests, it waits to hear back about specifically which ones have made progress, until all requests are driven to completion or the high-level timeout is exceeded.

### CODE REVIEW

**Comment by gitbuda** [comment]
@spacejam I think the `io::future` with the wait and fill notifiers is going to be very useful for me (the physical operators work) 😄 So, can you take that out of this PR and review + merge separately / before? I'm happy to do that / help, but you might have some additional extensions in mind. Also, not in a hurry. If this is close to the reviewable state, just shepherd it through 😄

**Comment by spacejam** [comment]
@gitbuda I think this should be merging quickly so I'll keep it together for now, but if there are surprises I'll split it out to get it merged by the end of the day

File: src/io/simulator/simulator.hpp [hunk 1]
```diff
@@ -49,5 +49,11 @@ class Simulator {
   }
 
   SimulatorStats Stats() { return simulator_handle_->Stats(); }
+
+  std::function<bool()> GetSimulatorTickClosure() {
+    std::shared_ptr<SimulatorHandle> handle_copy = simulator_handle_;
+    std::function<bool()> tick_closure = [handle_copy] { return handle_copy->MaybeTickSimulator(); };
```
Comment (gvolfing) [inline src/io/simulator/simulator.hpp:55]:
```suggestion
    std::function<bool()> tick_closure = [handle_copy = simulator_handle_] { return handle_copy->MaybeTickSimulator(); };
```

File: src/io/simulator/simulator_transport.hpp [hunk 2]
```diff
@@ -33,11 +34,13 @@ class SimulatorTransport {
       : simulator_handle_(simulator_handle), address_(address), rng_(std::mt19937{seed}) {}
 
   template <Message RequestT, Message ResponseT>
-  ResponseFuture<ResponseT> Request(Address to_address, Address from_address, RequestT request, Duration timeout) {
-    std::function<bool()> maybe_tick_simulator = [this] { return simulator_handle_->MaybeTickSimulator(); };
+  ResponseFuture<ResponseT> Request(Address to_address, Address from_address, RequestT request,
+                                    std::function<void()> notification, Duration timeout) {
+    std::shared_ptr<SimulatorHandle> handle_copy = simulator_handle_;
+    std::function<bool()> tick_simulator = [handle_copy] { return handle_copy->MaybeTickSimulator(); };
```
Comment (gvolfing) [inline src/io/simulator/simulator_transport.hpp:40]:
```suggestion
    std::function<bool()> tick_simulator = [handle_copy = simulator_handle_] { return handle_copy->MaybeTickSimulator(); };
```

**Review by gvolfing — APPROVED** [review]

File: src/query/v2/request_router.hpp [hunk 3]
```diff
@@ -370,25 +359,33 @@ class RequestRouter : public RequestRouterInterface {
...
+    auto responses = DriveReadResponses<msgs::GetPropertiesRequest, msgs::GetPropertiesResponse>(running_requests);
+
+    // post-process responses
+    std::vector<msgs::GetPropertiesResultRow> result_rows;
 
-    std::vector<msgs::GetPropertiesResultRow> result;
-    for (auto &res : responses) {
-      std::move(res.result_row.begin(), res.result_row.end(), std::back_inserter(result));
+    for (auto &&response : responses) {
+      std::move(response.result_row.begin(), response.result_row.end(), std::back_inserter(result_rows));
```
Comment (42jeremy) [inline src/query/v2/request_router.hpp:385]: I am wondering if std::make_move_iterator wouldn't be better here

File: src/io/rsm/rsm_client.hpp [hunk 4]
```diff
@@ -101,61 +93,63 @@ class RsmClient {
   ~RsmClient() = default;
 
   BasicResult<TimedOut, WriteResponseT> SendWriteRequest(WriteRequestT req) {
-    auto token = SendAsyncWriteRequest(req);
-    auto poll_result = AwaitAsyncWriteRequest(token);
+    Notifier notifier;
+    ReadinessToken readiness_token{0};
```
Comment (42jeremy) [inline src/io/rsm/rsm_client.hpp:97]:
```suggestion
    const ReadinessToken readiness_token{0};
```

File: src/io/rsm/rsm_client.hpp [hunk 5]
```diff
@@ -101,61 +93,63 @@ class RsmClient {
...
+      poll_result = AwaitAsyncWriteRequest(readiness_token);
     }
     return poll_result.value();
   }
 
   BasicResult<TimedOut, ReadResponseT> SendReadRequest(ReadRequestT req) {
-    auto token = SendAsyncReadRequest(req);
-    auto poll_result = AwaitAsyncReadRequest(token);
+    Notifier notifier;
+    ReadinessToken readiness_token{0};
```
Comment (42jeremy) [inline src/io/rsm/rsm_client.hpp:108]:
```suggestion
    const ReadinessToken readiness_token{0};
```

**Pull request merged**